### ISSUE TITLE
Occurences not emitted for `type` 

### ISSUE THREAD

**Issue by tgodzik** [scalameta/scalameta#2024]
```
object root {

  object impl // OK
  val f: impl.type => Unit = { // no occurence
    case _: impl.type => () // no occurence
  }
  f(impl) // OK
}
```

**Comment by tanishiking** [comment]
I took a quick look.

I'm not 100% sure, but it might be a problem with scalac? (`Ident(TermName("impl"))` in `SingletonTypeTree(Ident(TermName("impl"))` always have `NoSymbol`).

- `occurrences` in `TextDocument` are calculated by [finalOccurrences](https://github.com/scalameta/scalameta/blob/d8d6b9d9789f0c8f175260fd7002a5489a860704/semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala#L747-L757)
- `occurrences` and `mpatoccurrences` are gathered while [traversing tree](https://github.com/scalameta/scalameta/blob/d8d6b9d9789f0c8f175260fd7002a5489a860704/semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala#L199-L743)
  - on the other hand, the token positions are also indexed by [traversing tree](https://github.com/scalameta/scalameta/blob/d8d6b9d9789f0c8f175260fd7002a5489a860704/semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala#L79-L197), and position for `impl.type` looks successfully indexed.
- `Ident(TermName("impl"))` match [case gtree: g.RefTree](https://github.com/scalameta/scalameta/blob/d8d6b9d9789f0c8f175260fd7002a5489a860704/semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala#L400-L405) and add its symbol to `occurrences`.
- **The problem is**
  - While `impl` in `f(impl)` has an appropriate symbol and be added to occurrences,
    - `Ident(TermName("impl"))` in `Apply(Select(Ident(TermName("f")), TermName("apply")), List(Ident(TermName("impl"))))`
  - `impl` in `impl.type` always has `NoSymbol`
    -  `Ident(TermName("impl"))` in `SingletonTypeTree(Ident(TermName("impl"))`
- Then `impl.type` wouldn't be added to `occurrenes` because it's [uselessOccurrence](https://github.com/scalameta/scalameta/blob/d8d6b9d9789f0c8f175260fd7002a5489a860704/semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala#L246)

```scala
universe.showRaw(universe.reify {
  object impl
    val f: impl.type => Unit = {
     case _: impl.type => ()
   }
  f(impl)
}.tree )
```

```scala
Block(
  List(
    ModuleDef(
      Modifiers(),
      TermName("impl"), Template(List(Ident(TypeName("AnyRef"))), noSelfType, List(DefDef(Modifiers(), termNames.CONSTRUCTOR, List(), List(List()), TypeTree(), Block(List(Apply(Select(Super(This(typeNames.EMPTY), typeNames.EMPTY), termNames.CONSTRUCTOR), List())), Literal(Constant(()))))))
    ),
    ValDef(
      Modifiers(),
      TermName("f"),
      AppliedTypeTree(Ident(scala.Function1), List(SingletonTypeTree(Ident(TermName("impl"))), Ident(scala.Unit))),
      Function(
        List(ValDef(Modifiers(PARAM | SYNTHETIC), TermName("x0$1"), TypeTree(), EmptyTree)),
        Match(
          Ident(TermName("x0$1")),
          List(CaseDef(Typed(Ident(termNames.WILDCARD), SingletonTypeTree(Ident(TermName("impl")))), EmptyTree, Literal(Constant(()))))
        )
      )
    )
  ),
  Apply(Select(Ident(TermName("f")), TermName("apply")), List(Ident(TermName("impl"))))
)
```

**Comment by tgodzik** [comment]
Thanks for the thorough investigation! That was also my suspicion that we might need to fix it in scalac :/  which will require a bit more time.

**Comment by tanishiking** [comment]
Filed an issue on scala/bug https://github.com/scala/bug/issues/12296

**Comment by tanishiking** [comment]
The fix https://github.com/scala/scala/pull/9426 has merged and will be out in 2.12.14 and 2.13.5

**Comment by tgodzik** [comment]
> The fix [scala/scala#9426](https://github.com/scala/scala/pull/9426) has merged and will be out in 2.12.14 and 2.13.5

That is pretty awesome! Thanks!

**Comment by tgodzik** [comment]
LEt's maybe keep this open and add a test to the expect suite on
